source/io/module_eggShell.js

```diff
@@ -239,6 +246,257 @@
+            var typeRef = args[0].typeRef,
+                sizeOfComplex = Module.eggShell.typeTopAQSize(typeRef),
+                visitFn;
+            switch (sizeOfComplex) {
+            case 8:
+                visitFn = typeVisitor.visitComplexSingle;
+                break;
+            case 16:
+                visitFn = typeVisitor.visitComplexDouble;
+                break;
+            default:
+                throw new Error('Unexpected size for a Complex value');
+            }
+
+            validateVisitMethod(visitFn, 'visitComplex' + (sizeOfComplex === 8 ? 'Single' : 'Double'));
+            return visitFn.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitAnalogWaveform = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitAnalogWaveform, 'visitAnalogWaveform');
+            return typeVisitor.visitAnalogWaveform.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitTimestamp = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitTimestamp, 'visitTimestamp');
+            return typeVisitor.visitTimestamp.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitPath = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitPath, 'visitPath');
+            return typeVisitor.visitPath.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitArray = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitArray, 'visitArray');
+            return typeVisitor.visitArray.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitCluster = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitCluster, 'visitCluster');
+            return typeVisitor.visitCluster.apply(typeVisitor, args);
+        };
+
+        var typeDispatchFunctions = [
+            dispatchVisitBoolean,
+            dispatchVisitEnum,
+            dispatchVisitInteger,
+            dispatchVisitFloat,
+            dispatchVisitString,
+            dispatchVisitComplex,
+            dispatchVisitAnalogWaveform,
+            dispatchVisitTimestamp,
+            dispatchVisitPath,
+            dispatchVisitArray,
+            dispatchVisitCluster
+        ];
+
+        var registeredTypes = [
+            Module.eggShell.typeIsBoolean,
+            Module.eggShell.typeIsEnum,
+            Module.eggShell.typeIsInteger,
+            Module.eggShell.typeIsFloat,
+            Module.eggShell.typeIsString,
+            Module.eggShell.typeIsComplex,
+            Module.eggShell.typeIsAnalogWaveform,
+            Module.eggShell.typeIsTimestamp,
+            Module.eggShell.typeIsPath,
+            Module.eggShell.typeIsArray,
+            Module.eggShell.typeIsCluster
+        ];
+
+        Module.eggShell.reflectOnValueRef = publicAPI.eggShell.reflectOnValueRef = function (typeVisitor, valueRef) {
+            if (typeof valueRef !== 'object' || Array.isArray(valueRef)) {
```

Comment:
The Array check seems a bit arbitrary. For sanity checking that an object was passed
`typeof valueRef !== 'object' || valueRef === null`
is sufficient.

Same for below, if someone's prerogative is to do:
```
var myWackyVisisitor = [];
myWackyVisitor.visitArray = function () {};
```
then let 'em do it :P

Reply:
Yeah, it was just JS catching me off guard, didn't want to allow that, but ok. They'll be responsible for their wacky visitor.